src/databricks/labs/ucx/hive_metastore/grants.py

```diff
@@ -737,20 +752,22 @@ def __init__(
         self._group_manager = group_manager
         self._grant_loaders = grant_loaders
 
-    def apply(self, src: Table, uc_table_key: str) -> bool:
+    def apply(self, src: Catalog | Schema | Table, dst: Catalog | Schema | Table) -> bool:
```

Comment:
it looks like you need a duck-typed protocol (a.k.a. interface-on-read, coming from Go):

```
class WithFullName(typing.Protocol):
    def full_name(self) -> str: ...
```